### PR TITLE
Provide std::array replacement for kernel code

### PR DESCRIPTION
**Pull request by yuriykoch**
Provide replacement type for fixed-width array to use in kernel functions. Implicitly uses pointer arithmetic for C++ ranged-for.

### CODE REVIEW

**Comment by ProGTX** [comment]
Is there a specific reason for doing this? It generally looks good, I'm just curious what the motivation was.

File: util/array.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+  /** @brief Internal data storage;
+   *         public for aggregate initialization availability
+   */
+  value_type values[N];
+
+  reference operator[](size_type pos) {
+    return values[pos];
+  }
+  const_reference operator[](size_type pos) const {
```
Comment (ProGTX) [inline util/array.h:42]: All `const` qualified functions can be marked `constexpr` if targetting SYCL 1.2.1. For SYCL 2020 all functions can be `constexpr`.

File: util/array.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    return values[pos];
+  }
+
+  inline constexpr size_type size() const noexcept {
```
Comment (ProGTX) [inline util/array.h:46]:
```suggestion
  constexpr size_type size() const noexcept {
```

`constexpr` implies `inline` when used on functions.

File: util/array.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  const_iterator begin() const noexcept {
+    return &front();
+  }
+  iterator end() noexcept {
+    return &back();
+  }
+  const_iterator end() const noexcept {
+    return &back();
+  }
+};
```
Comment (ProGTX) [inline util/array.h:77]: I think we're missing a `data()` function

**Comment by bader** [comment]
> Is there a specific reason for doing this? It generally looks good, I'm just curious what the motivation was.

According to my understanding, it goal is provide exception-free implementation to avoid *potential* concerns regarding using std::array in SYCL device code.

File: util/array.h (hunk 3 above)
Comment (yuriykoch) [reply util/array.h:77]: Do we really need it on the device side? I mean, we can always do
`for (const auto& item: arrayInstance) {
  /* access item */
}`

File: util/array.h (hunk 1 above)
Comment (yuriykoch) [reply util/array.h]: Great catch, done

File: util/array.h (hunk 3 above)
Comment (ProGTX) [reply util/array.h:77]: Not having it would probably be the only difference to a standard `array`, and I presume we want a drop-in replacement.

**Review by ProGTX — APPROVED** [review]
I think there are more concerns that just exceptions when it comes to the general STL, and we need to address them at some point. This change looks good, though.

**Review by keryell — APPROVED** [review]
That looks good!

**Pull request merged**